### PR TITLE
Update Azure SQL Authentication to use Azure.Identity

### LINKED ISSUE #12: dbup-sqlserver uses deprecated Microsoft.Azure.Services.AppAuthentication
When updating our solutions packags we noticed that `dbup-sqlserver` has a dependency to `Microsoft.Azure.Services.AppAuthentication >= 1.6.2` which is **deprecated**. Suggesting an alternate implementation with regards to the suggestion in issue DbUp/DbUp#641.

### PR DESCRIPTION
**Pull request by evochriso**
# Checklist
- [X] I have read the [Contributing Guide](https://github.com/DbUp/DbUp/blob/master/CONTRIBUTING.md)
- [X] I have checked to ensure this does not introduce an unintended breaking changes
- [X] I have considered appropriate testing for my change

# Description
Replace deprecated package Microsoft.Azure.Services.AppAuthentication with Azure.Identity
Resolves #12

### CODE REVIEW

**Comment by droyad** [comment]
Thanks I'll pull this in when we drop support `System.Data.SqlClient` as this is also a breaking change.

**Comment by Bartleby2718** [comment]
@droyad If you don't mind educating a fellow engineer, would you be willing to explain why it'd be better to wait for another major version bump?

My understanding is that:
- The public API hasn't changed, so this change in and of itself is not a breaking change. 
- The new dependency [Azure.Identity doesn't have major version 2](https://www.nuget.org/packages/Azure.Identity#versions-body-tab), so I don't think consumers will be negatively affected by the new dependency.
- [The issue was first reported in July 2022](https://github.com/DbUp/DbUp/issues/641), so I'm sure there are people waiting for this.

(Hope I didn't miss anything obvious.)

**Comment by evochriso** [comment]
> Thanks I'll pull this in when we drop support `System.Data.SqlClient` (a breaking change).

@droyad this update shouldn't be a breaking change since it doesn't change behavior or API surface, just swaps a deprecated dependency for its replacement, and has no dependency on System.Data.SqlClient. Perhaps a minor version bump instead of wait for the next major release? 
As mentioned by @Bartleby2718 - the usage of the deprecated package is causing dbup-sqlserver to be flagged by security scanners and a higher priority issue than removing the System.Data.SqlClient package which is not yet deprecated (just needs version bump to resolve security vulnerability).

**Review by Swampen — COMMENTED** [review]
Just wanted to give my input on this after having written a _CustomConnectionManager_ to be able to use `Azure.Identity` earlier. We used `AzureCliCredential` for dev environments and `WorkloadIdentityCredential`/`ManagedIdentityCredential` for production environments to skip the loop of having to check all possible types of credentials.

File: src/dbup-sqlserver/AzureSqlConnectionManager.cs [hunk 1]
```diff
@@ -28,12 +29,11 @@ public AzureSqlConnectionManager(string connectionString, string resource)
         public AzureSqlConnectionManager(string connectionString, string resource, string tenantId, string azureAdInstance = "https://login.microsoftonline.com/")
             : base(new DelegateConnectionFactory((log, dbManager) =>
             {
+                var tokenProvider = new DefaultAzureCredential();
```
[the remark this answers is not in the document]
Comment (evochriso) [reply src/dbup-sqlserver/AzureSqlConnectionManager.cs:32]: @Swampen my goal with this PR was to resolve the usage of a deprecated Nuget package, without introducing a breaking change. In my opinion, using the `DefaultAzureCredential` here makes sense so we can be as flexible as possible, and overhead of checking all possible types of credentials seems negligible for a database upgrader that will just make one connection to perform the upgrade. Changing or adding the constructors for the `AzureSqlConnectionManager` class to accept other parameters such as `TokenCredential` would be a breaking change. 
I don't know what @droyad has planned for the next major release, but I agree we should change constructors in that release.

**Comment by droyad** [comment]
Alright how does that look?

If ok, could you give it a quick manual test and I'll merge it. I don't have AAD Integrated DB on hand.

**Comment by peymanr34** [comment]
Is it necessary to explicitly depend on `Azure.Identity`? The `Microsoft.Data.SqlClient` already depends on it.
 - or maybe you could just depend on [Azure.Core](https://www.nuget.org/packages/Azure.Core/)?

**Comment by droyad** [comment]
>Is it necessary to explicitly depend on Azure.Identity? 

It does not need it. I've removed it.

**Comment by evochriso** [comment]
@droyad consider the version of Azure.Identity - Microsoft.Data.SqlClient requires Azure.Identity 1.11.3 or greater, but that version of Azure.Identity has a vulnerability and should be bumped to 1.12.0

**Comment by peymanr34** [comment]
> @droyad consider the version of Azure.Identity - Microsoft.Data.SqlClient requires Azure.Identity 1.11.3 or greater, but that version of Azure.Identity has a vulnerability and should be bumped to 1.12.0

The `Microsoft.Data.SqlClient` dependencies are a mess, however I think you should do this in your own projects otherwise this repo will become the place to complain/ask to fix `Microsoft.Data.SqlClient`'s bad versioning practices.

**Comment by droyad** [comment]
The latest SqlClient references Identity `1.11.4` which doesn't have a vulnerability. We'll update SqlClient.

**Pull request merged**